### DISCUSSION TITLE
What is the range of shuffleID?

### DISCUSSION THREAD

**Discussion by awsnap** [RondeSC/Miller_Shuffle_Algo#5] [General]
I was wondering what the range of `shuffleID `could be. There are `Factorial(listSize)` different shuffles, yet glancing at the code (on Wikipedia) it seems there are at most `p1*p2*listSize` different shuffles generated by  `function PRIG()`. It looks like this is fine in practice, but if you wanted some really big permutations you could be in for a surprise. Is that right?

I don't know much about permutation generation.

**Comment by RondeSC** [comment]
Generally the shuffleID is 32bits unsigned, a very big number. The number of possible generated permutation is further limited by the degree it effects the shuffle via the r1-rx "random factors" along with the 'listsize'. This will realistically only result in a very small fraction of Factorial(listSize) permutations. I think mathematically it is the listSize to the power of the number of  "random factors" used, due to the effects of modulo math (%listSize). So if each 'shuffleID', r1, r2, r3 and r4 are used (as in MSA-d) to change 'si' and the shuffle mix, it would = listSize to the fifth. In which case, the 'shuffleID's 32bits is more limiting, it dictates the maximum number of possible generated permutations.
On a practical basis MSA-d appears to do as well as the Fisher-Yates algo in a test I performed to check the distribution frequency of 5040 (10 * 9 * 8 * 7) expected permutations within a given number of shuffles. Only the use of radiation data produces better statistics (which is not appropriate, because it does not produce a shuffle, instead a random mix). Study over and compare test results at https://docs.google.com/document/d/1tHUKb0QNdGcvMp39ofUcqJBImnKHFtMtx7zCVNSsNMU particularly for 'permu5k', 'deal2', & 'deal'  permutation tests,

**Comment by RondeSC** [comment]
I did a [million shuffles](https://docs.google.com/document/d/1ymoP43hNTPVr56KaZbcj-NIWiCtwsxwMTbzPXeS-cpE/edit?usp=sharing) with the algorithms and looked for repeated shuffle instances. With MSA-d I found >4700 repeats, with MSA-e I found ~50 repeats. With the randomly driven (non deterministic) MSA-b and Fisher-Yates I found <10. When I tested ½ as many shuffles I found ½ as many and when I tested twice as many I found twice as many repeats. 
I then tested sets of 10M shuffles; for MSA-E got 2300 repeats, MSA-Max got 750. The non-deterministic MSA-b only got 96, while for Fisher-Yate found 113 repeats (all after -240 which were due to random() & ShuffleID repeated values).
I think that MSA-e will give >4.2 billion unique shuffles out of the 4.29 billion 32b ShuffleIDs, while MSA-b will give about 99.999% unique shuffles and Fisher-Yates apparently slightly less.
Do you have a use case which calls for more generation capacity than 4 billion unique shuffles and/or less than 1 in 100,000 repeats ?

**Comment by RondeSC** [comment]
See previously:  README.md paragraph containing "**MSA_e results in Zero repeated shuffles**"
With current testing methodology, of a billion instances, using MSA-E, no repeated shuffle permutations are detected.
